From 59ce3053e35b60cf5c3e652f445eadb1440c5adf Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Thu, 1 Apr 2021 21:46:39 -0500
Subject: [PATCH 066/196] Make SIGHASH_FORKID and SIGHASH_BIP341 mutually
 exclusive (#123)

Drop the FORKID part when printing and parsing signatures.
This makes using the BIP341 sighash simpler, as signers now only have to specify e.g. ALL|BIP341 instead of ALL|BIP341|FORKID.

Also, SignatureHash now fails if the sigHashType is neither Legacy, BIP143 nor BIP341.

Set SIGHASH_BIP341 = 0x60 (was just 0x20), and add SIGHASH_LEGACY and SIGHASH_RESERVED.
Add SIGHASH_ALGORITHM_MASK to filter out the algorithm used for the sighash.
Add SIGHASH_UNUSED_MASK to filter out unused bits.
Add SIGHASH_BASE_TYPE_MASK to filter out the base type.
Add SigHashType::withAlgorithm and SigHashType::isLegacy, and SigHashType::isReserved.

SigHashType::withForkId and SigHashType::withBIP341 now no longer take a boolean.
With a false parameter, the resulting sighash algorithm would always be SIGHASH_LEGACY, which might not be desired.
This forces users to specify which algorithm to use.
---
 src/bitcoin-tx.cpp                    | 18 +++---
 src/core_read.cpp                     | 24 +++-----
 src/script/interpreter.cpp            | 13 +++--
 src/script/script_flags.h             |  2 +-
 src/script/sighashtype.h              | 82 +++++++++++++++++++++------
 src/test/data/script_tests.json       |  4 +-
 src/test/data/sighash_bip341.json     | 54 +++++++++---------
 src/test/schnorr_tests.cpp            |  7 ++-
 src/test/script_tests.cpp             | 10 +++-
 src/test/sigencoding_tests.cpp        | 16 ++++--
 src/test/sighash_bip341_tests.cpp     | 30 +++++-----
 src/test/sighash_tests.cpp            | 20 ++++---
 src/test/sighashtype_tests.cpp        | 74 +++++++++++++++---------
 test/util/data/bitcoin-util-test.json | 14 ++---
 14 files changed, 226 insertions(+), 142 deletions(-)

diff --git a/src/bitcoin-tx.cpp b/src/bitcoin-tx.cpp
index ce9beb119..6662f6e4b 100644
--- a/src/bitcoin-tx.cpp
+++ b/src/bitcoin-tx.cpp
@@ -543,15 +543,15 @@ static const struct {
      SIGHASH_NONE | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
     {"SINGLE|FORKID|ANYONECANPAY",
      SIGHASH_SINGLE | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
-    {"ALL|BIP341|FORKID", SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_FORKID},
-    {"NONE|BIP341|FORKID", SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_FORKID},
-    {"SINGLE|BIP341|FORKID", SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_FORKID},
-    {"ALL|BIP341|FORKID|ANYONECANPAY",
-     SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
-    {"NONE|BIP341|FORKID|ANYONECANPAY",
-     SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
-    {"SINGLE|BIP341|FORKID|ANYONECANPAY",
-     SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
+    {"ALL|BIP341", SIGHASH_ALL | SIGHASH_BIP341},
+    {"NONE|BIP341", SIGHASH_NONE | SIGHASH_BIP341},
+    {"SINGLE|BIP341", SIGHASH_SINGLE | SIGHASH_BIP341},
+    {"ALL|BIP341|ANYONECANPAY",
+     SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY},
+    {"NONE|BIP341|ANYONECANPAY",
+     SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY},
+    {"SINGLE|BIP341|ANYONECANPAY",
+     SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY},
 };
 
 static bool findSigHashFlags(SigHashType &sigHashType,
diff --git a/src/core_read.cpp b/src/core_read.cpp
index 85033f397..52c93d3b7 100644
--- a/src/core_read.cpp
+++ b/src/core_read.cpp
@@ -291,21 +291,15 @@ SigHashType ParseSighashString(const UniValue &sighash) {
             {"SINGLE|FORKID", SIGHASH_SINGLE | SIGHASH_FORKID},
             {"SINGLE|FORKID|ANYONECANPAY",
              SIGHASH_SINGLE | SIGHASH_FORKID | SIGHASH_ANYONECANPAY},
-            {"ALL|BIP341|FORKID",
-             SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_FORKID},
-            {"ALL|BIP341|FORKID|ANYONECANPAY", SIGHASH_ALL | SIGHASH_BIP341 |
-                                                   SIGHASH_FORKID |
-                                                   SIGHASH_ANYONECANPAY},
-            {"NONE|BIP341|FORKID",
-             SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_FORKID},
-            {"NONE|BIP341|FORKID|ANYONECANPAY", SIGHASH_NONE | SIGHASH_BIP341 |
-                                                    SIGHASH_FORKID |
-                                                    SIGHASH_ANYONECANPAY},
-            {"SINGLE|BIP341|FORKID",
-             SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_FORKID},
-            {"SINGLE|BIP341|FORKID|ANYONECANPAY",
-             SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_FORKID |
-                 SIGHASH_ANYONECANPAY},
+            {"ALL|BIP341", SIGHASH_ALL | SIGHASH_BIP341},
+            {"ALL|BIP341|ANYONECANPAY", SIGHASH_ALL | SIGHASH_BIP341 |
+                                            SIGHASH_ANYONECANPAY},
+            {"NONE|BIP341", SIGHASH_NONE | SIGHASH_BIP341},
+            {"NONE|BIP341|ANYONECANPAY", SIGHASH_NONE | SIGHASH_BIP341 |
+                                             SIGHASH_ANYONECANPAY},
+            {"SINGLE|BIP341", SIGHASH_SINGLE | SIGHASH_BIP341},
+            {"SINGLE|BIP341|ANYONECANPAY",
+             SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY},
         };
         std::string strHashType = sighash.get_str();
         const auto &it = map_sighash_values.find(strHashType);
diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index ee10bf0f2..45aa2cdeb 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -1874,18 +1874,19 @@ bool SignatureHash(uint256 &sighashOut,
         sigHashType = sigHashType.withForkValue(0xff0000 | newForkValue);
     }
 
-    if (!sigHashType.hasForkId() || !(flags & SCRIPT_ENABLE_SIGHASH_FORKID)) {
+    if (sigHashType.isLegacy() || !(flags & SCRIPT_ENABLE_SIGHASH_FORKID)) {
         return SignatureHashLegacy(sighashOut, scriptCode, txTo, nIn,
                                    sigHashType);
-    }
-
-    if (sigHashType.hasBIP341()) {
+    } else if (sigHashType.hasForkId()) {
+        return SignatureHashBIP143(sighashOut, scriptCode, txTo, nIn,
+                                   sigHashType, amount, cache);
+    } else if (sigHashType.hasBIP341()) {
         assert(cache);
         return SignatureHashBIP341(sighashOut, execdata, txTo, nIn, sigHashType,
                                    *cache);
     } else {
-        return SignatureHashBIP143(sighashOut, scriptCode, txTo, nIn,
-                                   sigHashType, amount, cache);
+        // reserved sigHashType 0x20
+        return false;
     }
 }
 
diff --git a/src/script/script_flags.h b/src/script/script_flags.h
index cae3be807..faa2c8a5e 100644
--- a/src/script/script_flags.h
+++ b/src/script/script_flags.h
@@ -35,7 +35,7 @@ enum {
     //
     SCRIPT_VERIFY_MINIMALIF = (1U << 13),
 
-    // Do we accept signature using SIGHASH_FORKID
+    // Do we accept signatures using SIGHASH_FORKID or SIGHASH_BIP341
     //
     SCRIPT_ENABLE_SIGHASH_FORKID = (1U << 16),
 
diff --git a/src/script/sighashtype.h b/src/script/sighashtype.h
index c355d22ec..023b06082 100644
--- a/src/script/sighashtype.h
+++ b/src/script/sighashtype.h
@@ -12,12 +12,30 @@
 
 /** Signature hash types/flags */
 enum {
+    //! Sign all outputs
     SIGHASH_ALL = 1,
+    //! Sign no outputs
     SIGHASH_NONE = 2,
+    //! Sign the output at the same index as the signed input
     SIGHASH_SINGLE = 3,
-    SIGHASH_BIP341 = 0x20,
-    SIGHASH_FORKID = 0x40,
+    //! Sign only this input, other inputs can be added or removed
     SIGHASH_ANYONECANPAY = 0x80,
+
+    //! Use the original legacy Bitcoin sighash algorim
+    SIGHASH_LEGACY = 0x00,
+    //! Invalid reserved sighash algorithm
+    SIGHASH_RESERVED = 0x20,
+    //! Use the BIP143 sighash algorithm
+    SIGHASH_FORKID = 0x40,
+    //! Use the BIP341 sighash algorithm
+    SIGHASH_BIP341 = 0x60,
+
+    //! Bits which specify which sighash algorithm to use
+    SIGHASH_ALGORITHM_MASK = 0x60,
+    //! Bits which currently are not used and must be 0
+    SIGHASH_UNUSED_MASK = 0x1c,
+    //! Bits encoding the output type (ALL, NONE, SINGLE) as well as unused bits
+    SIGHASH_BASE_TYPE_MASK = 0x1f,
 };
 
 /**
@@ -45,21 +63,26 @@ public:
     explicit SigHashType(uint32_t sigHashIn) : sigHash(sigHashIn) {}
 
     SigHashType withBaseType(BaseSigHashType baseSigHashType) const {
-        return SigHashType((sigHash & ~0x1f) | uint32_t(baseSigHashType));
+        return SigHashType((sigHash & ~SIGHASH_BASE_TYPE_MASK) |
+                           uint32_t(baseSigHashType));
     }
 
     SigHashType withForkValue(uint32_t forkId) const {
         return SigHashType((forkId << 8) | (sigHash & 0xff));
     }
 
-    SigHashType withForkId(bool forkId = true) const {
-        return SigHashType((sigHash & ~SIGHASH_FORKID) |
-                           (forkId ? SIGHASH_FORKID : 0));
+    SigHashType withForkId() const {
+        return SigHashType((sigHash & ~SIGHASH_ALGORITHM_MASK) |
+                           SIGHASH_FORKID);
     }
 
-    SigHashType withBIP341(bool bip341 = true) const {
-        return SigHashType((sigHash & ~SIGHASH_BIP341) |
-                           (bip341 ? SIGHASH_BIP341 : 0));
+    SigHashType withBIP341() const {
+        return SigHashType((sigHash & ~SIGHASH_ALGORITHM_MASK) |
+                           SIGHASH_BIP341);
+    }
+
+    SigHashType withAlgorithm(uint32_t algorithm) const {
+        return SigHashType((sigHash & ~SIGHASH_ALGORITHM_MASK) | algorithm);
     }
 
     SigHashType withAnyoneCanPay(bool anyoneCanPay = true) const {
@@ -68,21 +91,48 @@ public:
     }
 
     BaseSigHashType getBaseType() const {
-        return BaseSigHashType(sigHash & 0x1f);
+        return BaseSigHashType(sigHash & SIGHASH_BASE_TYPE_MASK);
     }
 
     uint32_t getForkValue() const { return sigHash >> 8; }
 
+    uint32_t getUnusedBits() const { return sigHash & SIGHASH_UNUSED_MASK; }
+
     bool isDefined() const {
-        auto baseType =
-            BaseSigHashType(sigHash & ~(SIGHASH_FORKID | SIGHASH_ANYONECANPAY));
-        return baseType >= BaseSigHashType::ALL &&
-               baseType <= BaseSigHashType::SINGLE;
+        BaseSigHashType baseType = getBaseType();
+        switch (baseType) {
+            case BaseSigHashType::ALL:
+            case BaseSigHashType::NONE:
+            case BaseSigHashType::SINGLE:
+                break;
+            default:
+                return false;
+        }
+        switch (sigHash & SIGHASH_ALGORITHM_MASK) {
+            case SIGHASH_LEGACY:
+            case SIGHASH_FORKID:
+                break;
+            default:
+                return false;
+        }
+        return true;
     }
 
-    bool hasForkId() const { return (sigHash & SIGHASH_FORKID) != 0; }
+    bool isLegacy() const {
+        return (sigHash & SIGHASH_ALGORITHM_MASK) == SIGHASH_LEGACY;
+    }
+
+    bool isReserved() const {
+        return (sigHash & SIGHASH_ALGORITHM_MASK) == SIGHASH_RESERVED;
+    }
 
-    bool hasBIP341() const { return (sigHash & SIGHASH_BIP341) != 0; }
+    bool hasForkId() const {
+        return (sigHash & SIGHASH_ALGORITHM_MASK) == SIGHASH_FORKID;
+    }
+
+    bool hasBIP341() const {
+        return (sigHash & SIGHASH_ALGORITHM_MASK) == SIGHASH_BIP341;
+    }
 
     bool hasAnyoneCanPay() const {
         return (sigHash & SIGHASH_ANYONECANPAY) != 0;
diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index 4a9c9dcf5..7709b90e6 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -2488,14 +2488,14 @@
     "P2PK with undefined hashtype"
 ],
 [
-    "0x47 0x3044022070fd4f47970fdb7814b79482d1132a7c3598437bef78f8458de6e5d66ce918fd02204e657846fee57cb6fc3cbaae6b85db8646cc351f96155c86387388c2e8a5a83a21 0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8",
+    "0x47 0x304402200581361d23e645be9e3efe63a9a2ac2e8dd0c70ba3ac8554c9befe060ad0b36202207d8172f1e259395834793d81b17e986f1e6131e4734969d2f4ae3a9c8bc4296521 0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8",
     "DUP HASH160 0x14 0x91b24bf9f5288532960ac687abb035127b1d28a5 EQUALVERIFY CHECKSIG",
     "",
     "SIG_HASHTYPE",
     "P2PKH with invalid sighashtype and STRICTENC"
 ],
 [
-    "0x47 0x30440220643b4c3979d9d2c41842d4843113c1b40ca1df85805c7c654cabb015208c23f702207d3675bf98419065ac9a24865b652551849a0b1e52fdc429cd150f5e7897627821 0x43 0x41048282263212c609d9ea2a6e3e172de238d8c39cabd5ac1ca10646e23fd5f5150811f8a8098557dfe45e8256e830b60ace62d613ac2f7b17bed31b6eaff6e26cafac",
+    "0x47 0x304402203097a64e2a79f0dfa5520e3069888c105f517c463a1ccbf9e3c3dcb796c17fc6022046857d27eec970662d20a77699928e9ab4fbabecc34fe92205587277091c72e621 0x43 0x41048282263212c609d9ea2a6e3e172de238d8c39cabd5ac1ca10646e23fd5f5150811f8a8098557dfe45e8256e830b60ace62d613ac2f7b17bed31b6eaff6e26cafac",
     "HASH160 0x14 0x49ba2f86705b5dcd48d93b750f03289db3b8ce21 EQUAL",
     "",
     "SIG_HASHTYPE",
diff --git a/src/test/data/sighash_bip341.json b/src/test/data/sighash_bip341.json
index f385e2a6e..8e00ac7e8 100644
--- a/src/test/data/sighash_bip341.json
+++ b/src/test/data/sighash_bip341.json
@@ -9,7 +9,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "ALL|BIP341|FORKID",
+        "ALL|BIP341",
         0,
         4294967295,
         "SIGHASH_FORKID",
@@ -35,7 +35,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "ALL|BIP341|FORKID",
+        "ALL|BIP341",
         1,
         256,
         "SIGHASH_FORKID",
@@ -61,7 +61,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "ALL|BIP341|FORKID|ANYONECANPAY",
+        "ALL|BIP341|ANYONECANPAY",
         0,
         4294967295,
         "SIGHASH_FORKID",
@@ -86,7 +86,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "ALL|BIP341|FORKID|ANYONECANPAY",
+        "ALL|BIP341|ANYONECANPAY",
         1,
         256,
         "SIGHASH_FORKID",
@@ -110,7 +110,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "NONE|BIP341|FORKID",
+        "NONE|BIP341",
         0,
         4294967295,
         "SIGHASH_FORKID",
@@ -135,7 +135,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "NONE|BIP341|FORKID",
+        "NONE|BIP341",
         1,
         256,
         "SIGHASH_FORKID",
@@ -159,7 +159,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "NONE|BIP341|FORKID|ANYONECANPAY",
+        "NONE|BIP341|ANYONECANPAY",
         0,
         4294967295,
         "SIGHASH_FORKID",
@@ -183,7 +183,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "NONE|BIP341|FORKID|ANYONECANPAY",
+        "NONE|BIP341|ANYONECANPAY",
         1,
         256,
         "SIGHASH_FORKID",
@@ -205,7 +205,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "SINGLE|BIP341|FORKID",
+        "SINGLE|BIP341",
         0,
         4294967295,
         "SIGHASH_FORKID",
@@ -229,7 +229,7 @@
          "01", "ff00000000000000", "0151",
          "00000000"],
         [[1111, "5"], [111111, "8"]],
-        "SINGLE|BIP341|FORKID",
+        "SINGLE|BIP341",
         1,
         4294967295,
         "SIGHASH_FORKID",
@@ -245,7 +245,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "SINGLE|BIP341|FORKID",
+        "SINGLE|BIP341",
         1,
         256,
         "SIGHASH_FORKID",
@@ -271,7 +271,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "SINGLE|BIP341|FORKID",
+        "SINGLE|BIP341",
         1,
         256,
         "SIGHASH_FORKID",
@@ -285,7 +285,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "SINGLE|BIP341|FORKID|ANYONECANPAY",
+        "SINGLE|BIP341|ANYONECANPAY",
         0,
         4294967295,
         "SIGHASH_FORKID",
@@ -310,7 +310,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "SINGLE|BIP341|FORKID|ANYONECANPAY",
+        "SINGLE|BIP341|ANYONECANPAY",
         1,
         256,
         "SIGHASH_FORKID",
@@ -335,7 +335,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "ALL|BIP341|FORKID",
+        "ALL|BIP341",
         0,
         4294967295,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -361,7 +361,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "ALL|BIP341|FORKID",
+        "ALL|BIP341",
         1,
         256,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -387,7 +387,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "ALL|BIP341|FORKID|ANYONECANPAY",
+        "ALL|BIP341|ANYONECANPAY",
         0,
         4294967295,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -412,7 +412,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "ALL|BIP341|FORKID|ANYONECANPAY",
+        "ALL|BIP341|ANYONECANPAY",
         1,
         256,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -436,7 +436,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "NONE|BIP341|FORKID",
+        "NONE|BIP341",
         0,
         4294967295,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -461,7 +461,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "NONE|BIP341|FORKID",
+        "NONE|BIP341",
         1,
         256,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -485,7 +485,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "NONE|BIP341|FORKID|ANYONECANPAY",
+        "NONE|BIP341|ANYONECANPAY",
         0,
         4294967295,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -509,7 +509,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "NONE|BIP341|FORKID|ANYONECANPAY",
+        "NONE|BIP341|ANYONECANPAY",
         1,
         256,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -531,7 +531,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "SINGLE|BIP341|FORKID",
+        "SINGLE|BIP341",
         0,
         4294967295,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -555,7 +555,7 @@
          "01", "ffff000000000000", "0151",
          "00000000"],
         [[1245, "3"],[1245, "3"]],
-        "SINGLE|BIP341|FORKID",
+        "SINGLE|BIP341",
         1,
         4294967295,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -571,7 +571,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "SINGLE|BIP341|FORKID",
+        "SINGLE|BIP341",
         1,
         256,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -596,7 +596,7 @@
                "ffffff0000000000", "0152",
          "00000000"],
         [[1245, "3"]],
-        "SINGLE|BIP341|FORKID|ANYONECANPAY",
+        "SINGLE|BIP341|ANYONECANPAY",
         0,
         4294967295,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
@@ -621,7 +621,7 @@
          "98765432"],
         [[65536, "OP_HASH160 0x14 0x5555555555555555555555555555555555555555 OP_EQUAL"],
          [65537, "OP_DUP OP_HASH160 0x14 0x2222222222222222222222222222222222222222 OP_EQUALVERIFY OP_CHECKSIG"]],
-        "SINGLE|BIP341|FORKID|ANYONECANPAY",
+        "SINGLE|BIP341|ANYONECANPAY",
         1,
         256,
         "SIGHASH_FORKID,REPLAY_PROTECTION",
diff --git a/src/test/schnorr_tests.cpp b/src/test/schnorr_tests.cpp
index d2654a81e..10b5e1e77 100644
--- a/src/test/schnorr_tests.cpp
+++ b/src/test/schnorr_tests.cpp
@@ -94,12 +94,13 @@ BOOST_AUTO_TEST_CASE(opcodes_random_flags) {
         uint32_t flags = lcg.next();
 
         const bool hasForkId = (flags & SCRIPT_ENABLE_SIGHASH_FORKID) != 0;
+        const SigHashType sigHashType = SigHashType().withAlgorithm(
+            hasForkId ? SIGHASH_FORKID : SIGHASH_LEGACY);
 
         // Prepare 65-byte transaction sigs with right hashtype byte.
-        valtype DER64_with_hashtype =
-            SignatureWithHashType(DER64, SigHashType().withForkId(hasForkId));
+        valtype DER64_with_hashtype = SignatureWithHashType(DER64, sigHashType);
         valtype Zero64_with_hashtype =
-            SignatureWithHashType(Zero64, SigHashType().withForkId(hasForkId));
+            SignatureWithHashType(Zero64, sigHashType);
 
         // Test CHECKSIG & CHECKDATASIG with he non-DER sig, which can fail from
         // encoding, otherwise upon verification.
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index cb5c784ef..6e4bdcb1c 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -342,9 +342,13 @@ public:
                  Amount amount = Amount::zero(),
                  uint32_t sigFlags = SCRIPT_ENABLE_SIGHASH_FORKID) {
         uint256 hash;
-        BOOST_CHECK(SignatureHash(
-            hash, std::optional(ScriptExecutionData(script)), script,
-            CTransaction(spendTx), 0, sigHashType, amount, nullptr, sigFlags));
+        // Only sign if not reserved sighash, sign 0x0000...00 otherwise
+        if (!sigHashType.isReserved()) {
+            BOOST_CHECK(
+                SignatureHash(hash, std::optional(ScriptExecutionData(script)),
+                              script, CTransaction(spendTx), 0, sigHashType,
+                              amount, nullptr, sigFlags));
+        }
         std::vector<uint8_t> vchSig = DoSignECDSA(key, hash, lenR, lenS);
         vchSig.push_back(static_cast<uint8_t>(sigHashType.getRawSigHashType()));
         DoPush(vchSig);
diff --git a/src/test/sigencoding_tests.cpp b/src/test/sigencoding_tests.cpp
index 3516edb32..276fae32f 100644
--- a/src/test/sigencoding_tests.cpp
+++ b/src/test/sigencoding_tests.cpp
@@ -38,7 +38,8 @@ static void CheckSignatureEncodingWithSigHashType(const valtype &vchSig,
 
     for (const SigHashType &baseSigHash : baseSigHashes) {
         // Check the signature with the proper forkid flag.
-        SigHashType sigHash = baseSigHash.withForkId(hasForkId);
+        SigHashType sigHash = baseSigHash.withAlgorithm(
+            hasForkId ? SIGHASH_FORKID : SIGHASH_LEGACY);
         valtype validSig = SignatureWithHashType(vchSig, sigHash);
         BOOST_CHECK(CheckTransactionSignatureEncoding(validSig, flags, &err));
         BOOST_CHECK_EQUAL(!is64, CheckTransactionECDSASignatureEncoding(
@@ -67,7 +68,8 @@ static void CheckSignatureEncodingWithSigHashType(const valtype &vchSig,
         }
 
         // If we check strict encoding, then invalid forkid is an error.
-        SigHashType invalidSigHash = baseSigHash.withForkId(!hasForkId);
+        SigHashType invalidSigHash =
+            baseSigHash.withAlgorithm(hasForkId ? SIGHASH_LEGACY : SIGHASH_FORKID);
         valtype invalidSig = SignatureWithHashType(vchSig, invalidSigHash);
 
         BOOST_CHECK(
@@ -372,10 +374,12 @@ BOOST_AUTO_TEST_CASE(checkschnorr_test) {
         const bool hasForkId = (flags & SCRIPT_ENABLE_SIGHASH_FORKID) != 0;
 
         ScriptError err = ScriptError::OK;
-        valtype DER65_hb =
-            SignatureWithHashType(DER64, SigHashType().withForkId(hasForkId));
-        valtype Zero65_hb =
-            SignatureWithHashType(Zero64, SigHashType().withForkId(hasForkId));
+        valtype DER65_hb = SignatureWithHashType(
+            DER64, SigHashType().withAlgorithm(hasForkId ? SIGHASH_FORKID
+                                                         : SIGHASH_LEGACY));
+        valtype Zero65_hb = SignatureWithHashType(
+            Zero64, SigHashType().withAlgorithm(hasForkId ? SIGHASH_FORKID
+                                                          : SIGHASH_LEGACY));
 
         BOOST_CHECK(CheckDataSignatureEncoding(DER64, flags, &err));
         BOOST_CHECK(CheckTransactionSignatureEncoding(DER65_hb, flags, &err));
diff --git a/src/test/sighash_bip341_tests.cpp b/src/test/sighash_bip341_tests.cpp
index 64a648247..1dbf1a2ca 100644
--- a/src/test/sighash_bip341_tests.cpp
+++ b/src/test/sighash_bip341_tests.cpp
@@ -184,14 +184,17 @@ BOOST_AUTO_TEST_CASE(script_execution_data) {
 
 // Make sure BIP341 can't be used yet on the chain.
 BOOST_AUTO_TEST_CASE(bip341_not_enabled_yet) {
-    SigHashType sigHashType = SigHashType().withForkId().withBIP341();
+    SigHashType sigHashType = SigHashType().withBIP341();
     BOOST_CHECK(!sigHashType.isDefined());
-    BOOST_CHECK(!SigHashType(0x20).isDefined()); // SIGHASH_BUG
-    BOOST_CHECK(!sigHashType.withForkId(false).isDefined());
+    BOOST_CHECK(!SigHashType(SIGHASH_RESERVED).isDefined());
     BOOST_CHECK(!sigHashType.withAnyoneCanPay().isDefined());
-    BOOST_CHECK(!sigHashType.withForkId(false).withAnyoneCanPay().isDefined());
     BOOST_CHECK(!sigHashType.withBaseType(BaseSigHashType::NONE).isDefined());
     BOOST_CHECK(!sigHashType.withBaseType(BaseSigHashType::SINGLE).isDefined());
+    // withForkId resets the other bit in SIGHASH_TYPE_MASK
+    BOOST_CHECK(sigHashType.withAlgorithm(SIGHASH_LEGACY).isDefined());
+    BOOST_CHECK(sigHashType.withAlgorithm(SIGHASH_LEGACY)
+                    .withAnyoneCanPay()
+                    .isDefined());
 }
 
 BOOST_AUTO_TEST_CASE(bip341_invalid_hash_type) {
@@ -204,20 +207,21 @@ BOOST_AUTO_TEST_CASE(bip341_invalid_hash_type) {
 
     MMIXLinearCongruentialGenerator lcg;
     for (uint32_t sig_bits = 0; sig_bits <= 0xff; ++sig_bits) {
-        uint32_t hash_type = (lcg.next() << 8) | sig_bits;
-        bool is_valid = true;
-        if (((hash_type & SIGHASH_BIP341) && (hash_type & SIGHASH_FORKID)) &&
-            (!(hash_type & 0x03) || (hash_type & 0x1c))) {
-            // hash_type 0 is invalid, any undefined bits are invalid
+        SigHashType sig_hash_type{(lcg.next() << 8) | sig_bits};
+        bool is_valid = !sig_hash_type.isReserved();
+        // Invalid bits make SignatureHash using BIP341 fail
+        if (sig_hash_type.hasBIP341() &&
+            (!(sig_bits & 0x03) || (sig_hash_type.getUnusedBits()))) {
+            // hash_type 0 is invalid, any usused bits are invalid
             is_valid = false;
         }
         const bool success = SignatureHash(
-            sighash, execdata, CScript(), tx, 0, SigHashType(hash_type),
-            Amount::zero(), &txdata, SCRIPT_ENABLE_SIGHASH_FORKID);
+            sighash, execdata, CScript(), tx, 0, sig_hash_type, Amount::zero(),
+            &txdata, SCRIPT_ENABLE_SIGHASH_FORKID);
         BOOST_CHECK_EQUAL(is_valid, success);
         if (is_valid != success) {
-            BOOST_ERROR("Unexpected result for hash type: " << std::hex
-                                                            << hash_type);
+            BOOST_ERROR("Unexpected result for hash type: "
+                        << std::hex << sig_hash_type.getRawSigHashType());
         }
     }
 }
diff --git a/src/test/sighash_tests.cpp b/src/test/sighash_tests.cpp
index 6a42ba1c3..d8c46f1a2 100644
--- a/src/test/sighash_tests.cpp
+++ b/src/test/sighash_tests.cpp
@@ -132,17 +132,18 @@ BOOST_AUTO_TEST_CASE(sighash_test) {
     int nRandomTests = 1000;
     for (int i = 0; i < nRandomTests; i++) {
         uint32_t nHashType = InsecureRand32();
-        if (nHashType & SIGHASH_BIP341) {
-            nHashType &= ~uint32_t(0x1c); // clear invalid bits
-            nHashType |= SIGHASH_FORKID;  // prevent illegal flag combination
-            if (!(nHashType & 0x1f)) {    // base type not set
+        // BIP341 must have 0x40 set and valid base type
+        if (nHashType & SIGHASH_RESERVED) {
+            nHashType &= ~uint32_t(SIGHASH_UNUSED_MASK); // clear invalid bits
+            nHashType |= SIGHASH_FORKID; // prevent illegal flag combination
+            if (!(nHashType & SIGHASH_BASE_TYPE_MASK)) { // base type not set
                 nHashType |= SIGHASH_ALL;
             }
         }
         SigHashType sigHashType(nHashType);
 
         CMutableTransaction txTo;
-        RandomTransaction(txTo, (nHashType & 0x1f) == SIGHASH_SINGLE);
+        RandomTransaction(txTo, (nHashType & SIGHASH_BASE_TYPE_MASK) == SIGHASH_SINGLE);
         CScript scriptCode;
         RandomScript(scriptCode);
         const std::optional<ScriptExecutionData> execdata{
@@ -166,10 +167,15 @@ BOOST_AUTO_TEST_CASE(sighash_test) {
                                   CTransaction(txTo), nIn, sigHashType,
                                   Amount::zero(), &txdata));
         if (sigHashType.hasForkId()) {
-            BOOST_CHECK(nHashType & SIGHASH_FORKID);
+            BOOST_CHECK_EQUAL(nHashType & SIGHASH_ALGORITHM_MASK,
+                              SIGHASH_FORKID);
+            BOOST_CHECK(shreg != shref);
+        } else if (sigHashType.hasBIP341()) {
+            BOOST_CHECK_EQUAL(nHashType & SIGHASH_ALGORITHM_MASK,
+                              SIGHASH_BIP341);
             BOOST_CHECK(shreg != shref);
         } else {
-            BOOST_CHECK((nHashType & SIGHASH_FORKID) == 0);
+            BOOST_CHECK_EQUAL(nHashType & SIGHASH_FORKID, 0);
             BOOST_CHECK(shreg == shref);
         }
 
diff --git a/src/test/sighashtype_tests.cpp b/src/test/sighashtype_tests.cpp
index 0bc7d3f4f..e0b04a84c 100644
--- a/src/test/sighashtype_tests.cpp
+++ b/src/test/sighashtype_tests.cpp
@@ -16,68 +16,83 @@ BOOST_FIXTURE_TEST_SUITE(sighashtype_tests, BasicTestingSetup)
 
 static void CheckSigHashType(SigHashType t, BaseSigHashType baseType,
                              bool isDefined, uint32_t forkValue, bool hasForkId,
+                             bool hasBIP341, uint32_t unusedBits,
                              bool hasAnyoneCanPay) {
     BOOST_CHECK(t.getBaseType() == baseType);
     BOOST_CHECK_EQUAL(t.isDefined(), isDefined);
     BOOST_CHECK_EQUAL(t.getForkValue(), forkValue);
+    BOOST_CHECK_EQUAL(t.getUnusedBits(), unusedBits);
     BOOST_CHECK_EQUAL(t.hasForkId(), hasForkId);
+    BOOST_CHECK_EQUAL(t.hasBIP341(), hasBIP341);
     BOOST_CHECK_EQUAL(t.hasAnyoneCanPay(), hasAnyoneCanPay);
 }
 
 BOOST_AUTO_TEST_CASE(sighash_construction_test) {
     // Check default values.
-    CheckSigHashType(SigHashType(), BaseSigHashType::ALL, true, 0, false,
-                     false);
+    CheckSigHashType(SigHashType(), BaseSigHashType::ALL, true, 0, false, false,
+                     0, false);
 
     // Check all possible permutations.
     std::set<BaseSigHashType> baseTypes{
         BaseSigHashType::UNSUPPORTED, BaseSigHashType::ALL,
         BaseSigHashType::NONE, BaseSigHashType::SINGLE};
     std::set<uint32_t> forkValues{0, 1, 0x123456, 0xfedcba, 0xffffff};
-    std::set<bool> forkIdFlagValues{false, true};
+    std::set<uint32_t> algorithmValues{SIGHASH_LEGACY, SIGHASH_FORKID,
+                                       SIGHASH_BIP341};
     std::set<bool> anyoneCanPayFlagValues{false, true};
 
     for (BaseSigHashType baseType : baseTypes) {
         for (uint32_t forkValue : forkValues) {
-            for (bool hasForkId : forkIdFlagValues) {
+            for (uint32_t algorithm : algorithmValues) {
                 for (bool hasAnyoneCanPay : anyoneCanPayFlagValues) {
-                    const SigHashType t =
-                        SigHashType()
-                            .withBaseType(baseType)
-                            .withForkValue(forkValue)
-                            .withForkId(hasForkId)
-                            .withAnyoneCanPay(hasAnyoneCanPay);
+                    SigHashType t = SigHashType()
+                                        .withBaseType(baseType)
+                                        .withForkValue(forkValue)
+                                        .withAlgorithm(algorithm)
+                                        .withAnyoneCanPay(hasAnyoneCanPay);
+                    bool hasForkId = algorithm == SIGHASH_FORKID;
+                    bool hasBIP341 = algorithm == SIGHASH_BIP341;
 
                     bool isDefined = baseType != BaseSigHashType::UNSUPPORTED;
-                    CheckSigHashType(t, baseType, isDefined, forkValue,
-                                     hasForkId, hasAnyoneCanPay);
+                    CheckSigHashType(t, baseType, isDefined && !hasBIP341,
+                                     forkValue, hasForkId, hasBIP341, 0,
+                                     hasAnyoneCanPay);
 
                     // Also check all possible alterations.
-                    CheckSigHashType(t.withForkId(hasForkId), baseType,
-                                     isDefined, forkValue, hasForkId,
-                                     hasAnyoneCanPay);
-                    CheckSigHashType(t.withForkId(!hasForkId), baseType,
-                                     isDefined, forkValue, !hasForkId,
-                                     hasAnyoneCanPay);
-                    CheckSigHashType(t.withAnyoneCanPay(hasAnyoneCanPay),
+                    CheckSigHashType(t.withAlgorithm(hasForkId
+                                                         ? SIGHASH_FORKID
+                                                         : SIGHASH_LEGACY),
                                      baseType, isDefined, forkValue, hasForkId,
+                                     false, 0, hasAnyoneCanPay);
+                    CheckSigHashType(t.withAlgorithm(!hasForkId
+                                                         ? SIGHASH_FORKID
+                                                         : SIGHASH_LEGACY),
+                                     baseType, isDefined, forkValue, !hasForkId,
+                                     false, 0, hasAnyoneCanPay);
+                    CheckSigHashType(t.withAnyoneCanPay(hasAnyoneCanPay),
+                                     baseType, isDefined && !hasBIP341,
+                                     forkValue, hasForkId, hasBIP341, 0,
                                      hasAnyoneCanPay);
                     CheckSigHashType(t.withAnyoneCanPay(!hasAnyoneCanPay),
-                                     baseType, isDefined, forkValue, hasForkId,
+                                     baseType, isDefined && !hasBIP341,
+                                     forkValue, hasForkId, hasBIP341, 0,
                                      !hasAnyoneCanPay);
 
                     for (BaseSigHashType newBaseType : baseTypes) {
                         bool isNewDefined =
-                            newBaseType != BaseSigHashType::UNSUPPORTED;
+                            newBaseType != BaseSigHashType::UNSUPPORTED &&
+                            !hasBIP341;
                         CheckSigHashType(t.withBaseType(newBaseType),
                                          newBaseType, isNewDefined, forkValue,
-                                         hasForkId, hasAnyoneCanPay);
+                                         hasForkId, hasBIP341, 0,
+                                         hasAnyoneCanPay);
                     }
 
                     for (uint32_t newForkValue : forkValues) {
                         CheckSigHashType(t.withForkValue(newForkValue),
-                                         baseType, isDefined, newForkValue,
-                                         hasForkId, hasAnyoneCanPay);
+                                         baseType, isDefined && !hasBIP341,
+                                         newForkValue, hasForkId, hasBIP341, 0,
+                                         hasAnyoneCanPay);
                     }
                 }
             }
@@ -93,8 +108,12 @@ BOOST_AUTO_TEST_CASE(sighash_serialization_test) {
         for (uint32_t forkValue : forkValues) {
             uint32_t rawType = sigHashType | (forkValue << 8);
 
-            uint32_t baseType = rawType & 0x1f;
-            bool hasForkId = (rawType & SIGHASH_FORKID) != 0;
+            uint32_t baseType = rawType & SIGHASH_BASE_TYPE_MASK;
+            uint32_t unused = rawType & SIGHASH_UNUSED_MASK;
+            bool hasForkId =
+                (rawType & SIGHASH_ALGORITHM_MASK) == SIGHASH_FORKID;
+            bool hasBIP341 =
+                (rawType & SIGHASH_ALGORITHM_MASK) == SIGHASH_BIP341;
             bool hasAnyoneCanPay = (rawType & SIGHASH_ANYONECANPAY) != 0;
 
             uint32_t noflag =
@@ -105,7 +124,8 @@ BOOST_AUTO_TEST_CASE(sighash_serialization_test) {
 
             // Check deserialization.
             CheckSigHashType(tbase, BaseSigHashType(baseType), isDefined,
-                             forkValue, hasForkId, hasAnyoneCanPay);
+                             forkValue, hasForkId, hasBIP341, unused,
+                             hasAnyoneCanPay);
 
             // Check raw value.
             BOOST_CHECK_EQUAL(tbase.getRawSigHashType(), rawType);
diff --git a/test/util/data/bitcoin-util-test.json b/test/util/data/bitcoin-util-test.json
index 66aaf4758..5f98c707b 100644
--- a/test/util/data/bitcoin-util-test.json
+++ b/test/util/data/bitcoin-util-test.json
@@ -370,7 +370,7 @@
      "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
      "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
      "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
-     "sign=ALL|BIP341|FORKID"],
+     "sign=ALL|BIP341"],
     "output_cmp": "txcreatesignbip341v1-all.json",
     "description": "Creates a new v1 transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash (output in json)"
   },
@@ -382,7 +382,7 @@
      "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
      "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
      "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
-     "sign=ALL|BIP341|FORKID|ANYONECANPAY"],
+     "sign=ALL|BIP341|ANYONECANPAY"],
     "output_cmp": "txcreatesignbip341v2-all-anyonecanpay.json",
     "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with ANYONECANPAY (output in json)"
   },
@@ -394,7 +394,7 @@
      "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
      "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
      "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
-     "sign=ALL|BIP341|FORKID"],
+     "sign=ALL|BIP341"],
     "output_cmp": "txcreatesignbip341v2-all.json",
     "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash (output in json)"
   },
@@ -406,7 +406,7 @@
      "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
      "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
      "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
-     "sign=NONE|BIP341|FORKID|ANYONECANPAY"],
+     "sign=NONE|BIP341|ANYONECANPAY"],
     "output_cmp": "txcreatesignbip341v2-none-anyonecanpay.json",
     "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with NONE and ANYONECANPAY (output in json)"
   },
@@ -418,7 +418,7 @@
      "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
      "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
      "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
-     "sign=NONE|BIP341|FORKID"],
+     "sign=NONE|BIP341"],
     "output_cmp": "txcreatesignbip341v2-none.json",
     "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with NONE (output in json)"
   },
@@ -430,7 +430,7 @@
      "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
      "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
      "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
-     "sign=SINGLE|BIP341|FORKID|ANYONECANPAY"],
+     "sign=SINGLE|BIP341|ANYONECANPAY"],
     "output_cmp": "txcreatesignbip341v2-single-anyonecanpay.json",
     "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with SINGLE and ANYONECANPAY (output in json)"
   },
@@ -442,7 +442,7 @@
      "set=privatekeys:[\"KwDiBf89QgGbjEhKnhXJuH7LrciVrZi3qYjgd9M7rFU74NMTptX4\"]",
      "set=prevtxs:[{\"txid\":\"4d49a71ec9da436f71ec4ee231d04f292a29cd316f598bb7068feccabdc59485\",\"vout\":0,\"scriptPubKey\":\"76a91406afd46bcdfd22ef94ac122aa11f241244a37ecc88ac\"}]",
      "outaddr=0.001:qqr2l4rteh7j9mu54sfz4gglysfyfgm7esufu9gq2x",
-     "sign=SINGLE|BIP341|FORKID"],
+     "sign=SINGLE|BIP341"],
     "output_cmp": "txcreatesignbip341v2-single.json",
     "description": "Creates a new transaction with a single input and a single output, and then signs the transaction using the BIP341 sighash with SINGLE (output in json)"
   },
-- 
2.39.5 (Apple Git-154)

